### ISSUE TITLE
[ALL CHARTS] publish Prometheus community charts to OCI registry

### ISSUE THREAD

**Issue by scottrigby** [prometheus-community/helm-charts#2454]
### Is your feature request related to a problem ?

This January Helm [completed and released full OCI support](https://helm.sh/blog/storing-charts-in-oci/), moving OCI from experimental to a fully supported feature. Since then users have been able to install Helm charts completely from OCI rather than only from Helm HTTP repos.

However, Prometheus community Helm chart users are not able to pull or install Prometheus charts using Helm's OCI registry support, because they are not yet published to a registry.

Additionally, OCI method can reduce memory usage for users installing charts using an agent in the cloud rather than running the Helm client on their local machine, such as Flux. This is especially true of Helm repos with many charts, such as this Prometheus community charts repo, which contains over 30 charts. In a similarly large Helm repo, [Bitnami has recently removed all Helm chart package versions older than six months](https://github.com/bitnami/charts/issues/10539) from their repository because the index file has become too large to support all past versions.

### Describe the solution you'd like.

Publish all existing Prometheus community chart package versions to an OCI registry.

- GitHub Packages on ghcr.io would be the easiest way to set this up
- Community [Flux charts publish to ghcr.io](https://github.com/orgs/fluxcd-community/packages?repo_name=helm-charts). We have automation set up for this, which we can emulate
- As an admin of this Prometheus community charts repo, and also as a Helm and Flux maintainer, I can help set this up here as well
- The additional OCI support here will not affect the existing HTTP Helm repo support users currently depend on. It will just give those and other users another option to use OCI as well or as a replacement

### Describe alternatives you've considered.

We could publish to some other OCI registry, or not publish at all.
If we don't at all, I think the most likely scenario is someone else will fork and set it up on their own. 
I believe it will be better to support this here, so all packages are connected to one Git repo.

### Additional context.

I don't imagine any objection here. Mainly opening this issue to track the work and for transparency. Will keep status updated here 💖

**Comment by monotek** [comment]
No objections as long as we run both in parallel at least for a while :)

**Comment by scottrigby** [comment]
@monotek  👋 

> No objections as long as we run both in parallel at least for a while :)

Agreed 👍 

> We migth use something like: https://github.com/appany/helm-oci-chart-releaser

had a look at that action. It requires adding config for each chart, which I think is not ideal. But even less ideal is it requires updating action config every time you update a chart version - individually inputting each new chart version of each chart in the workflow config. I think that won't fit the needs of this repo.

> Just need to check how we can add all existing charts, but should be scriptable.

Yes adding all existing Prometheus community Helm package versions can definitely be scripted. We did a similar thing for the now archived helm/stable and helm/incubator repos using a tool I wrote (https://github.com/scottrigby/helm-adopt-package-history). That tool currently only supports HTTP Helm repos, but I think we can pretty easily add OCI push support as well. Or really whatever tool works in a foolproof and verifiable way.

> Not sure if the old release timestamp can be used though.

This can be done for images with OCI metadata. Will have to look into the best way to do this with OCI artifacts according to distribution spec. Let's just make sure this is considered before copying previous chart package versions to the OCI registry.

**Comment by monotek** [comment]
> had a look at that action. It requires adding config for each chart, which I think is not ideal. But even less ideal is it requires updating action config every time you update a chart version - individually inputting each new chart version of each chart in the workflow config. I think that won't fit the needs of this repo.

My favoutrite solution would also be that the ct / chart releaser action would be extendet to support oci releases too.
Not sure if it's worth to wait. There are at least issues available discussing it (https://github.com/helm/chart-releaser-action/issues/107 & https://github.com/helm/chart-releaser/issues/183).

**Comment by jkroepke** [comment]
Does someone not that all current releases are failing to ghcr.io with 403 forbidden?

**Comment by onedr0p** [comment]
Yea, it looks like the new versions of OCI charts have stopped being released.

**Comment by geowalrus4gh** [comment]
In this world of OCI registries(for Helm charts too), we expect this issue to be closed soon.

**Comment by monotek** [comment]
If we find a reliable way to use OCI releases i'm open to activate it again:
https://github.com/prometheus-community/helm-charts/blob/main/.github/workflows/release.yaml#L44C7-L70
I would keep it deactivated if CI is failing constantly beacause of it again.

**Comment by jkroepke** [comment]
What are the issues? Sadly, the build logs from #2719 are not available. 

Can we separate OCI Push and cosign? For example adding cosign after having stable OCI pushes?

Would it be possible to add `continue-on-error` to observe the errors while not blocking the CI?

**Comment by onedr0p** [comment]
@monotek maybe this could be reactivated again since the logs are gone?

**Comment by jkroepke** [comment]
And put the `continue-on-error` on the action that we are able to observe the error while not blocking us.

**Comment by monotek** [comment]
pr welcome

**Comment by jkroepke** [comment]
@monotek #3433 opened

**Comment by monotek** [comment]
Push failed with 403 :/
https://github.com/prometheus-community/helm-charts/actions/runs/5088616788/jobs/9145245233

**Comment by jkroepke** [comment]
@monotek 

Can you check, if github actions is allowed to publish into the packages? You have to go to https://github.com/prometheus-community/helm-charts/pkgs/container/charts%2Fprometheus-rabbitmq-exporter and then Packages settings on the right:

<img width="1020" alt="image" src="https://github.com/prometheus-community/helm-charts/assets/1560587/ea2d31e5-ee2d-4ed3-a45d-facfbf56c7ff">

On my repository, the permissions was granted automatically for actions, because the actions initially create the package on ghcr. This behavior is documented [here](https://docs.github.com/en/packages/managing-github-packages-using-github-actions-workflows/publishing-and-installing-a-package-with-github-actions#default-permissions-and-access-settings-for-packages-modified-through-workflows). Could it be possible that some of GHCR repositories are initially created manually for testing reasons? In that case, it could be possible that actions doesn't habe any access. 

Is it possible to nuke all the current package chart repositories and let them create by actions only? That should just works, if I'm reading the docs correctly.

**Comment by monotek** [comment]
Hmm... most of the packages were missing the setting.
Did it manually now.
Hopefully it will work automatically for new charts.

**Comment by jkroepke** [comment]
Lets observse this for three weeks. If it dont have any issues, i would to change the level from warning to error.

**Comment by monotek** [comment]
Meanwhile i'll write a script which pushes all existing charts to ghcr....

**Comment by onedr0p** [comment]
@monotek if/when you do that I can out using them. No rush, I doubt it's going to be a fun task to work on :)

**Comment by monotek** [comment]
It's already uploading....

**Comment by jkroepke** [comment]
Please validate, in case you create new charts, check the settings that I mentions.

**Comment by monotek** [comment]
prometheus-consul-exporter package seems to have some issue i can't fix.
the settings button is missing so i guess it's some permission issue.

@scottrigby 
could you have a look please?

https://github.com/prometheus-community/helm-charts/pkgs/container/charts%2Fprometheus-consul-exporter

**Comment by onedr0p** [comment]
Would it be possible to delete the package and try it again?

**Comment by monotek** [comment]
I can't. Need organisation admin like @scottrigby to do it.

**Comment by jkroepke** [comment]
@monotek No clue, but scottrigby seems not like an org admin to me.

Looking at https://github.com/orgs/prometheus-community/people?query=role%3Aowner @SuperQ can help us here.

**Comment by monotek** [comment]
Not sure if it should be the default.
Maybe add both ways?

OCI of prometheus-consul-exporter is still broken.

**Comment by onedr0p** [comment]
Is there anyway (slack, discord, IRC etc..) to directly reach out to @SuperQ to get him notified of the issue with that chart?

**Comment by SuperQ** [comment]
Sorry, I've been busy and not had time to look at this issue. I'll read over it today and see what I can do.

**Comment by SuperQ** [comment]
Ok, I think I fixed the `prometheus-consul-exporter` package access/permissions. Let me know if that fixes the issue and if there are any other packages to fix.

**Comment by monotek** [comment]
@SuperQ 
thanks a lot, it works :)
consul exporter packages are available now too in the oci registry. 

@jkroepke
we could move forward to adjust the charts readmes....

**Comment by jkroepke** [comment]
Can it all in one PR or each chart separtly?

**Comment by monotek** [comment]
The action will not work for multiple charts, so it has to be done separately.

But it might be enough to just add it to the repo readme for now:

https://github.com/prometheus-community/helm-charts#readme

**Comment by lilioid** [comment]
I just discovered this issue when looking for OCI releases of helm charts. 
I would greatly appreciate it, if those OCI release were documented either in the README or in individual charts installation instructions. 
Additionally, the artifacthub pages should also mention the possibility of installing a chart via OCI (see [example which doesn't do that](https://artifacthub.io/packages/helm/prometheus-community/prometheus-nats-exporter?modal=install)).

I can provide a PR for the documentation but am not sure how to get artifacthub to pick up on it. Maybe someone with higher access rights can just configure it.

**Comment by jkroepke** [comment]
Do you interesting into submit a PR?

**Comment by monotek** [comment]
I guess if you don't want to touch every chart's readme in a separate pr, you could just add some info to the [repo readme](https://github.com/prometheus-community/helm-charts/blob/main/README.md) that the oci artifacts of all charts can be found here: https://github.com/orgs/prometheus-community/packages?repo_name=helm-charts

**Comment by lilioid** [comment]
I do think, it should be mentioned in the installation instructions that are already part of each charts README.

So I don't mind creating a PR for every package if you don't mind the PR spam ^^

**Comment by monotek** [comment]
If you don't mind to do it... Thanks! :)

I'll do a pr for the repo readme -> https://github.com/prometheus-community/helm-charts/pull/5886

**Comment by lilioid** [comment]
Do we want to close this issue now?

**Comment by monotek** [comment]
Many thanks @lilioid :)

All done. Closing....
